src/safepoints.rs

```diff
@@ -0,0 +1,138 @@
+/// these are the frame pointer and stack pointer registers respectively. To
+/// avoid calculating this during a GC pause, we convert all offsets to be
+/// from an SP upfront.
+fn as_sp_offset(offset: &LocOffset) -> SPO {
+    match offset {
+        LocOffset::I32(ref o) => SPO(*o as u32),
+        _ => panic!("Offset must be signed"),
+    }
+}
+
+fn gen_safepoint_roots(stackmap: SMRec) -> SafepointRoots {
+    // The first 2 locations are uninteresting, however, they should be constants.
+    let mut idx = 2;
+    debug_assert_eq!(
+        mem::discriminant(&stackmap.locs[0].kind),
+        mem::discriminant(&LocKind::Constant)
+    );
+    debug_assert_eq!(
+        mem::discriminant(&stackmap.locs[1].kind),
+        mem::discriminant(&LocKind::Constant)
+    );
+
+    // The 3rd location specifies the number of de-opt locations. De-opt
+    // params are not interesting to us, so we skip over them
+    let num_deopts = {
+        match stackmap.locs[idx].offset {
+            LocOffset::U32(c) => c,
+            _ => panic!("Constants must be u32"),
+        }
+    };
+    idx += (num_deopts as usize) + 1;
+
+    // The remaining indices in the loc vector should all for GC pointers.
+    // There should be 2 pointers in this list for each GC pointer in the
+    // IR: a base pointer; and a derived pointer.
+    //
+    // We check that the number of remaining values is even
+    debug_assert!((stackmap.locs.len() - idx) % 2 == 0);
+
+    let mut offsets = Vec::new();
+
+    let mut gc_ptrs = stackmap.locs.iter().skip(idx);
+
+    while let Some(base) = gc_ptrs.next() {
+        let derived = gc_ptrs.next().unwrap();
+        match base.kind {
+            LocKind::Register => {
+                eprintln!("UNIMPLEMENTED: Skipping Registers for now");
+            }
+            LocKind::Indirect => match derived.kind {
+                LocKind::Indirect => {
+                    if base.offset == derived.offset {
+                        offsets.push(PtrSlot::Base(as_sp_offset(&base.offset)))
+                    } else {
+                        offsets.push(PtrSlot::Derived(
+                            as_sp_offset(&base.offset),
+                            as_sp_offset(&derived.offset),
+                        ))
+                    }
+                }
+                _ => unimplemented!(),
+            },
+            _ => eprintln!("UNIMPLEMENTED: Skipping over value"),
+        }
+    }
+
+    SafepointRoots {
+        registers: Vec::new(),
+        stack_offsets: offsets,
+    }
+}
+
+/// Generates a safepoint table which can be used during GC to lookup
+/// information about where pointers reside in a program.
+///
+/// This function will parse the .llvm_stackmap section of the given ELF file
+/// and generate an efficient hashmap -- keyed by a function's return address --
+/// which can be queried by the collector.
+pub fn gen_safepoint_table(path: &Path) -> HashMap<ReturnAddress, SafepointRoots> {
```

Comment:
In general, you might as well use `AsRef` with `Path` so you'd have something like:

```
pub fn gen_safepoint_table<P: AsRef<Path>>(path: P) -> ...
```

because it makes the API a little easier to use in some situations.

Reply:
Nice, I'll do this